src/dbnode/storage/namespace.go

```diff
@@ -1589,3 +1591,155 @@ func (n *dbNamespace) FlushState(shardID uint32, blockStart time.Time) (fileOpSt
+		n.metrics.writeAggData.ReportError(n.nowFn().Sub(callStart))
+		return errNamespaceNotBootstrapped
+	}
+	nsCtx := n.nsContextWithRLock()
+	targetShards := n.OwnedShards()
+	n.RUnlock()
+
+	// Note: Cold writes must be enabled for Large Tiles to work.
+	if !n.nopts.ColdWritesEnabled() {
+		n.metrics.writeAggData.ReportError(n.nowFn().Sub(callStart))
+		return errColdWritesDisabled
+	}
+
+	sourceNsOpts := sourceNs.StorageOptions()
+	reader, err := fs.NewReader(sourceNsOpts.BytesPool(), sourceNsOpts.CommitLogOptions().FilesystemOptions())
+	if err != nil {
+		n.metrics.writeAggData.ReportError(n.nowFn().Sub(callStart))
+		return err
+	}
+
+	wOpts := series.WriteOptions{
+		TruncateType: n.opts.TruncateType(),
+		SchemaDesc:   nsCtx.Schema,
+	}
+
+	resources, err := newColdFlushReuseableResources(n.opts)
+	if err != nil {
+		n.metrics.writeAggData.ReportError(n.nowFn().Sub(callStart))
+		return err
+	}
+
+	// NB(bodu): Deferred targetShard cold flushes so that we can ensure that cold flush index data is
+	// persisted before persisting TSDB data to ensure crash consistency.
+	multiErr := xerrors.NewMultiError()
+	for _, targetShard := range targetShards {
+		sourceShard, _, err := sourceNs.readableShardAt(targetShard.ID())
+		if err != nil {
+			detailedErr := fmt.Errorf("no matching shard in source namespace %s: %v", sourceNs.ID(), err)
+			multiErr = multiErr.Add(detailedErr)
+			continue
+		}
+		err = targetShard.AggregateTiles(ctx, reader, sourceNs.ID(), sourceShard, opts, wOpts)
+		if err != nil {
+			detailedErr := fmt.Errorf("shard %d aggregation failed: %v", targetShard.ID(), err)
+			multiErr = multiErr.Add(detailedErr)
+			continue
+		}
+
+		multiErr = n.coldFlushSingleShard(nsCtx, targetShard, pm, resources, multiErr)
+	}
+
+	res := multiErr.FinalError()
+	n.metrics.writeAggData.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	return res
+}
+
+func (n *dbNamespace) coldFlushSingleShard(
+	nsCtx namespace.Context,
+	shard databaseShard,
+	pm persist.Manager,
+	resources coldFlushReuseableResources,
+	multiErr xerrors.MultiError,
+) xerrors.MultiError {
+	// NB(rartoul): This value can be used for emitting metrics, but should not be used
+	// for business logic.
+	callStart := n.nowFn()
+
+	// NB(bodu): The in-mem index will lag behind the TSDB in terms of new series writes. For a period of
+	// time between when we rotate out the active cold mutable index segments (happens here) and when
+	// we actually cold flush the data to disk we will be making writes to the newly active mutable seg.
+	// This means that some series can live doubly in-mem and loaded from disk until the next cold flush
+	// where they will be evicted from the in-mem index.
+	var (
+		onColdFlushDone OnColdFlushDone
+		err             error
+	)
+	if n.reverseIndex != nil {
+		onColdFlushDone, err = n.reverseIndex.ColdFlush([]databaseShard{shard})
+		if err != nil {
+			n.metrics.writeAggData.ReportError(n.nowFn().Sub(callStart))
+			return multiErr.Add(
```

Comment:
nit: returning multi error here and below is a little awkward since these are single errors.

Reply:
Hmm, looks like this method (`coldFlushSingleShard`) takes a `multiErr` argument to accumulate errors across calls. Is that correct, @gediminasgu ?

Reply:
@notbdu as I see there could be shard ColdFlush, namespace onColdFlush/shard ColdFlush.Done and flushPersist.DoneFlush error.  As I see we can get here 3 errors, that's why I used here multiErr in place of a single err. Please let me know if we see it in the same way.

Reply:
Hmm, so I haven't seen any prior art where we return a `MultiError` from a fn call. I think the usage is prob fine. But usually we'd return single errors here and then below we'd accumulate errors in the `MultiError` and return `multiErr.FinalError()` at the end which is consistent with the go way of returning an `error` type from fns.

Reply:
Makes sense. I'll change it.